spd/eval.py

```diff
@@ -755,6 +756,165 @@ def compute(self) -> Mapping[str, float]:
+        self.component_activation_counts: dict[str, Float[Tensor, " C"]] = {
+            module_name: torch.zeros(model.C, device=self.device)
+            for module_name in model.components
+        }
+
+    def _load_reference_model(self) -> ComponentModel:
+        """Load the reference model from wandb or local path"""
+        if self.reference_model is None:
+            from spd.models.component_model import ComponentModel
+
+            assert self.reference_run_path is not None, (
+                "reference_run_path should not be None at this point"
+            )
+            self.reference_model = ComponentModel.from_pretrained(self.reference_run_path)
+
+            if torch.cuda.is_available():
+                self.reference_model.to("cuda")
+                self.reference_model.eval()
+                self.reference_model.requires_grad_(False)
+
+        return self.reference_model
+
+    def _compute_subcomponent_geometric_similarities(
+        self, activation_densities: dict[str, Float[Tensor, " C"]]
+    ) -> dict[str, float]:
+        """Compute mean max cosine similarity between subcomponent rank-one matrices"""
+        reference_model = self._load_reference_model()
+        similarities = {}
+
+        # Iterate through all component layers in both models
+        for layer_name in self.model.components:
+            if layer_name not in reference_model.components:
+                logger.warning(f"Layer {layer_name} not found in reference model, skipping")
+                continue
+
+            current_components = self.model.components[layer_name]
+            reference_components = reference_model.components[layer_name]
+
+            # Verify component counts match
+            if current_components.C != reference_components.C:
+                logger.warning(
+                    f"Component count mismatch for {layer_name}: {current_components.C} vs {reference_components.C}"
+                )
+                continue
+
+            # Extract U and V matrices
+            C = current_components.C
+            current_U = current_components.U  # Shape: [C, d_out]
+            current_V = current_components.V  # Shape: [d_in, C]
+            ref_U = reference_components.U
+            ref_V = reference_components.V
+
+            # Throw away components that are not active enough in the current model
+            density_threshold = self.kwargs.get("density_threshold", 0.0)
+            C_alive = sum(activation_densities[layer_name] > density_threshold)
+            if C_alive == 0:
+                logger.warning(
+                    f"\n WARNING:No components are active enough in {layer_name} for density threshold {density_threshold}. Geometric similarity comparison failed to run. \n"
+                )
+                continue
+            current_V = current_V[:, activation_densities[layer_name] > density_threshold]
+            current_U = current_U[activation_densities[layer_name] > density_threshold]
```

Comment:
I'd use a different name to current_V and current_U since this is a different thing to the original current_V and current_U